### DISCUSSION TITLE
Ebyte E22 modules[Board]:

### DISCUSSION THREAD

**Discussion by NancyDeville** [meshtastic/firmware#3948] [Troubleshooting]
### SOC

ESP32

### Lora IC

Ebyte E22-900M30S / Ebyte E22-900M33S

### Product Link

Aliexpress

### Description

I've been trying to achieve full transmit power with these Ebyte modules for around a year and so far none of my builds have performed to give full power output, they all top out at 22db representing the standard SX126x,
I have tried everything at this point in terms of wiring options of TX_EN/DIO2 also tried all sorts of regulators and capacitors in power supply, people on discord have sent variant.h files trying different things and nothing has worked,
I'm building firmware DIY hydra, V1 /V1.1,
After hours of online research I think the issue is with the settings in radio lib that in particular the register codes, txco settings, regulator mode, PA_ CONFIG and PA_Ramptime, 
Reading other projects online we also need to change max power settings in TX_PARMS from 22 to 30 for the E22-900M30S, note that also the E22-900M33S has different requirements and only needs 9db drive, and seems to have some different things in setup, I think again registers are different, 
Myself I'm more into hardware and I'm not really knowledgeable when it comes to writing code so I'm hoping someone can have a look at this and fill in the blanks, from what I can see it looks like we need a setup specifically for the ebyte modules that include PA and LNA, treating them totally as a standard SX1262x is not running them to their full potential as legal limits and power restrictions are written into the meshtastic radiolib

**Comment by Vanguard4893** [comment]
Hi mate - I've used these modules also - it might not be what you think, but hardware pins. When RXEN is Low on the E22 you need to pull TXEN high and vice versa. This can be done easily with a single NPN transistor.

Cheers

**Reply by NancyDeville** [reply]
I have tried both modes i.e manual and auto in regards to wiring, I've not done a setup using a transistor but the reason being that I don't see the need, I am getting the correct control behaviour on the pins

**Reply by Vanguard4893** [reply]
I see,

If you're getting the correct control behaviour on the RXEN/TXEN pins then it's a little strange - on my boards with the same module, 22dBm output on the radio IC gives the full 30dBm on the output given the gain from the PA.

For reference I am running with a PicoW microcontroller, on the firmware intended for the Pico/Waveshare setup, and have replicated this electrical layout exactly and get the expected results.

**Reply by NancyDeville** [reply]
I have a picow here with meshtastic on already, if you can provide schematic for connections Ill give it a try, did u make any changes to variant.h?

**Reply by NancyDeville** [reply]
I say schematic, pin list would be fine

**Reply by NancyDeville** [reply]
Are they using npn transistor on the waveshare ?

**Reply by NancyDeville** [reply]
I cant find a schematic for the transistor circuit, I can see they do use one, even still it would be nice if the DIY targets actually worked properly with the ebyte modules, I would like to hear some confirmation that the registers are correct for the E22 series, PA_CONFIG also says Todo, I know from research the ebytes are fussy and trigger happy when it comes to the protection circuit clamping the PA, I've seen a few projects talk of fixes for this

**Reply by NancyDeville** [reply]
This is what I'm looking at : see code below 
This I found online while looking into my issues, I can't find where we are doing this in meshtastic, the link to the feed where the discussion was is here, it seems to have sorted the issue here :

https://github.com/beegee-tokyo/SX126x-Arduino/issues/12

:	void SX126xSetTxParams(int8_t power, RadioRampTimes_t rampTime)
	{
		uint8_t buf[2];

		if (SX126xGetPaSelect(0) == SX1261)
		{
			if (power == 15)
			{
				SX126xSetPaConfig(0x06, 0x00, 0x01, 0x01);
			}
			else
			{
				SX126xSetPaConfig(0x04, 0x00, 0x01, 0x01);
			}
			if (power >= 14)
			{
				power = 14;
			}
			else if (power < -17)
			{
				power = -17;
			}
			SX126xWriteRegister(REG_OCP, 0x18); // current max is 80 mA for the whole device
		}
		else // sx1262
		{
			// WORKAROUND - Better Resistance of the SX1262 Tx to Antenna Mismatch, see DS_SX1261-2_V1.2 datasheet chapter 15.2
			// RegTxClampConfig = @address 0x08D8
			SX126xWriteRegister(0x08D8, SX126xReadRegister(0x08D8) | (0x0F << 1));
			// WORKAROUND END

			SX126xSetPaConfig(0x04, 0x07, 0x00, 0x01);
			if (power > 30) // was 22
			{
				power = 30; // was 22
			}
			else if (power < -9)
			{
				power = -9;
			}
			SX126xWriteRegister(REG_OCP, 0x38); // current max 160mA for the whole device
		}
		buf[0] = power;
		buf[1] = (uint8_t)rampTime;
		SX126xWriteCommand(RADIO_SET_TXPARAMS, buf, 2);
	}

**Reply by zagarolo** [reply]
> Hi mate - I've used these modules also - it might not be what you think, but hardware pins. When RXEN is Low on the E22 you need to pull TXEN high and vice versa. This can be done easily with a single NPN transistor.
> 
> Cheers

Hi help my schematic connect npn transistor please , enable Tx_en and Rx_en npn transistor  in module    Ebyte E07-433M20S

**Reply by zagarolo** [reply]
https://www.cdebyte.com/products/E07-433M20S/2#Pin

**Reply by NancyDeville** [reply]
I have done lots of testing for a year and tried both methods of connecting the E22 modules, I'm getting expected behaviour on pins but not expected output and high current pull for the low output

**Reply by NancyDeville** [reply]
Having eliminated the connection options from testing as being the issue and trying multiple regulators to power the module as well as multiple capacitors on the module power lines I really feel this is held back in software hence I'm digging into the code much as I know but I need someone with extensive knowledge of the firmware to look again at how these modules are being controlled in meshtastic, other projects as I say have had and overcome the issue by changing bits in code

**Reply by NancyDeville** [reply]
I really feel at the root of the problem is how restrictions are set to comply with use by general public in different parts of the world and knowing how to unlock this if you are licensed, I don't know but I think we might need to look again at the register codes meshtastic is setting up the module to use

**Reply by zagarolo** [reply]
in flipper zero board working example no firmware in module ability Pa hardware 

https://flipperaddons.com/product/e07-433m20s/

https://flipperaddons.com/product/sub-ghz-e07-433m20s/


no code is hardware example in flipper zero ability Pa 20 db only hardware

**Reply by NancyDeville** [reply]
The E07 is a very different module to the E22-900M30S it's based on cc1101 max 20db

**Comment by NancyDeville** [comment]
Ok so ebyte got back to me, they don't know meshtastic and can't offer any support, 
I've done a hell of alot of reading on this issue and the more I read up on the ebyte E22 full power output issue the more it seems to now be a meshtastic issue, 
Other projects that have had this issue have overcome the problem now buy changing things in the code,

I have managed to find the bits in the code I was looking for regarding power limits and regions and max tx power and pa ramp time etc, and I've spent about 40 hours now trying different things with no luck,

Me being into hardware and not doing much code I really wanted the team here to have another look into this issue and maybe write a profile specifically for the E23 modules, I say modules plural as we now have the E22-900M33S which has different setup requirements than the E22-900M30S ( 33 only needs 9db drive) ( this need to be at radio_begin otherwise we are potentially going to damage modules during channel changes as I've noted this also updates output power as part of the ch settings)

I have noted EBYTE_E22 in the code but I don't know just how much work has been done to fully support it let alone the issue I mentioned with the latest module so that would need it's own definition 

I don't feel moving this from and issue as I reported it to a discussion helps, there has been endless discussion and there is obviously an issue here

Anyway I'm hoping we can resolve this issue and get on with using these modules to their full potential

**Reply by madeofstown** [reply]
I'm not sure if you've seen this post in the Meshtastic Discourse forums. Maybe it could be some help to you. https://meshtastic.discourse.group/t/meshtasticd-on-pi-zero-2-and-1-watt-915mhz-ebyte-lora-rf-module/11459

**Comment by fifieldt** [comment]
Since this post was written, many more devices have used the E22 module. Refer to https://github.com/meshtastic/firmware/blob/efca2b5849d1636914f0da6fbfdab2cad9d746c0/variants/diy/nrf52_promicro_diy_tcxo/variant.h#L163 for the latest settings.

**Reply by K3v-68** [reply]
I am not into code at all, in simple steps, what shall I do to make my 33S version work as expected ?

**Reply by fifieldt** [reply]
This will depend heavily on your hardware. Assuming it is the same as the DIY promicro spec: https://github.com/meshtastic/firmware/tree/efca2b5849d1636914f0da6fbfdab2cad9d746c0/variants/diy/nrf52_promicro_diy_tcxo

Then,
Get the firmware-nrf52840 firmware release zip from the https://github.com/meshtastic/firmware/releases page for the release you want to run
Extract the firmware-nrf52_promicro_diy_tcxo-*.uf2 file
flash it.

If your hardware does not match the DIY promicro spec, you could brick your device.

Suggest asking for live assistance on discord.

**Reply by K3v-68** [reply]
Thanks for the quick answer, I'll dive a bit more into it. Kev

**Reply by itsKV** [reply]
Hello Kevin, 
Any success?

**Comment by OBakurov** [comment]
if E33 was connected and the code from E30 was used, then it is highly likely that it burned out, because for it it is necessary to reduce the output power.
